src/ioctls/vcpu.rs

```diff
@@ -2375,6 +2535,132 @@ mod tests {
+            SyncReg::VcpuEvents,
+        ];
+
+        for reg in &sync_regs {
+            vcpu.set_sync_dirty_reg(*reg);
+            assert_eq!(vcpu.kvm_run_ptr.as_mut_ref().kvm_dirty_regs, *reg as u64);
+            vcpu.clear_sync_dirty_reg(*reg);
+            assert_eq!(vcpu.kvm_run_ptr.as_mut_ref().kvm_dirty_regs, 0);
+        }
+
+        // Test that multiple dirty SyncRegs can be set at the same time
+        vcpu.set_sync_dirty_reg(SyncReg::Register);
+        vcpu.set_sync_dirty_reg(SyncReg::SystemRegister);
+        vcpu.set_sync_dirty_reg(SyncReg::VcpuEvents);
+        assert_eq!(
+            vcpu.kvm_run_ptr.as_mut_ref().kvm_dirty_regs,
+            SyncReg::Register as u64 | SyncReg::SystemRegister as u64 | SyncReg::VcpuEvents as u64
+        );
+    }
+
+    #[cfg(target_arch = "x86_64")]
+    #[test]
+    fn test_sync_regs_with_run() {
+        use std::io::Write;
+
+        let kvm = Kvm::new().unwrap();
+        let vm = kvm.create_vm().unwrap();
+        // This example is based on https://lwn.net/Articles/658511/
+        #[rustfmt::skip]
+        let code = [
+            0xff, 0xc0, /* inc eax */
+            0xf4, /* hlt */
+        ];
+
+        let mem_size = 0x4000;
+        let load_addr = mmap_anonymous(mem_size);
+        let guest_addr: u64 = 0x1000;
+        let slot: u32 = 0;
+        let mem_region = kvm_userspace_memory_region {
+            slot,
+            guest_phys_addr: guest_addr,
+            memory_size: mem_size as u64,
+            userspace_addr: load_addr as u64,
+            flags: KVM_MEM_LOG_DIRTY_PAGES,
+        };
+        unsafe {
+            vm.set_user_memory_region(mem_region).unwrap();
+        }
+
+        unsafe {
+            // Get a mutable slice of `mem_size` from `load_addr`.
+            // This is safe because we mapped it before.
+            let mut slice = std::slice::from_raw_parts_mut(load_addr, mem_size);
+            slice.write_all(&code).unwrap();
+        }
+
+        let mut vcpu = vm.create_vcpu(0).unwrap();
+
+        let orig_sregs = vcpu.get_sregs().unwrap();
+
+        let mut sync_regs = vcpu.sync_regs_mut();
+
+        // Initialize the sregs in sync_regs to be the original sregs
+        sync_regs.sregs = orig_sregs;
+        sync_regs.sregs.cs.base = 0;
+        sync_regs.sregs.cs.selector = 0;
+
+        // Set up the guest to attempt to `inc rax`
+        sync_regs.regs.rip = guest_addr;
+        sync_regs.regs.rax = 0x8000;
+        sync_regs.regs.rflags = 2;
+
+        // Initialize the sync_reg flags
+        vcpu.set_sync_valid_reg(SyncReg::Register);
+        vcpu.set_sync_valid_reg(SyncReg::SystemRegister);
+        vcpu.set_sync_valid_reg(SyncReg::VcpuEvents);
+        vcpu.set_sync_dirty_reg(SyncReg::Register);
+        vcpu.set_sync_dirty_reg(SyncReg::SystemRegister);
+        vcpu.set_sync_dirty_reg(SyncReg::VcpuEvents);
+
+        // Ensure sync reg bits are available for use
+        assert!(kvm.check_extension(Cap::SyncRegs));
```

Comment:
This way the test is still going to fail on some machines. I propose we run this test only if we have the capability, similar to how it is done [here](https://github.com/rust-vmm/kvm-ioctls/blob/main/src/ioctls/vcpu.rs#L1833):
```rust
if kvm.check_extension(Cap::SyncRegs) {
       let vm = kvm.create_vm().unwrap();
        // This example is based on https://lwn.net/Articles/658511/
       ...
        assert_eq!(sync_regs.regs.rax, 0x8001);
}
```

I just noticed there are a few places where we are going with the `assert` approach, but we should probably update them.